common/scala/src/main/scala/whisk/core/connector/MessageConsumer.scala

```diff
@@ -163,6 +171,26 @@ class MessageFeed(description: String,
         goto(FillingPipeline)
       } else stay
 
+    case Event(GracefulShutdown, _) =>
+      goto(GracefulShutdownDrain)
+
+    case _ => stay
+  }
+
+  when(GracefulShutdownDrain) {
+    case Event(Processed, _) =>
+      updateHandlerCapacity()
+      sendOutstandingMessages()
+      stay
+
+    case Event(FillCompleted(messages), _) =>
```

Comment:
Still concerned about a race condition here. There might be a chance that the `Busy` event returns `false` while `fillPipeline()` is running and before the `FillCompleted` event is triggered.

Reply:
right - if filling, you need to wait.
you'll also need to prevent a further fill.

Reply:
Hmm, do we know if events are handled synchronous or asynchronous in the Akka FSM?